### ISSUE TITLE
[RFE] Retry mechanism for HTTP 500 error when pulling ostree

### ISSUE THREAD

**Issue by oglok** [ostreedev/ostree#2747]
@mangelajo and I have been experiencing issues with pulling ostrees hosted in S3 buckets. This is framed in the context of the creation of ISO images, and the whole process would be interrupted if the request to pull one filez throws an HTTP 500 error.

We would like to extend the existing retrying mechanism to cover this type of http error.

Feedback would be much appreciated.

Thanks in advance.

**Comment by mangelajo** [comment]
I have also found this on rhel4edge when triggering an update from the edge management console, i.e.: 
![image](https://user-images.githubusercontent.com/1176434/198263411-98e27a59-538e-4be9-9db6-73992c6f2325.png)

**Comment by mangelajo** [comment]
@cgwalters @lucab Is the repo fetching also rust native in rpm-ostree, or is it consumed from here? If it is, I guess we should open a parallel RFE in rpm-ostree.

**Comment by oglok** [comment]
ok, let me push a PR to tackle this. Thanks for the info

**Comment by lucab** [comment]
Maybe I'm missing some details, but how did you deduce this happened in reaction to an HTTP 500 response?
The logs you pasted seem to point instead to `CURLE_RECV_ERROR`, which should be some kind of problem with the underlying network/transport instead, I think. At that point `libcurl` is possibly having issues with your whole networking I/O in general (i.e.possibly affecting DNS, TLS, etc.), not just with a HTTP `GET` to retry.

**Comment by oglok** [comment]
> Maybe I'm missing some details, but how did you deduce this happened in reaction to an HTTP 500 response? The logs you pasted seem to point instead to `CURLE_RECV_ERROR`, which should be some kind of problem with the underlying network/transport instead, I think. At that point `libcurl` is possibly having issues with your whole networking I/O in general (i.e.possibly affecting DNS, TLS, etc.), not just with a HTTP `GET` to retry.

```
**CloudWatch Logs Insights**  
region: us-east-1  
log-group-names: workers-aoc-production  
start-time: 2022-10-25T09:31:10.551Z  
end-time: 2022-10-25T09:34:42.316Z  
query-string:
```
fields message
| filter(host like "worker-aoc-production-i-0519acbbbd2aa4eb7")
| sort @timestamp desc
| limit 3000
```
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|                                                                                                                                                  message                                                                                                                                                  |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <info>  [1666690432.9629] dhcp4 (eth0): state changed new lease, address=172.31.43.22                                                                                                                                                                                                                     |
| <info>  [1666690432.9635] policy: set-hostname: current hostname was changed outside NetworkManager: 'worker-aoc-production-i-0519acbbbd2aa4eb7'                                                                                                                                                          |
| <info>  [1666690432.9638] policy: set-hostname: current hostname was changed outside NetworkManager: 'worker-aoc-production-i-0519acbbbd2aa4eb7'                                                                                                                                                          |
| time="2022-10-25T09:33:47Z" level=info msg="Instance protection removed"                                                                                                                                                                                                                                  |
| time="2022-10-25T09:33:47Z" level=warning msg="Job '77f48f72-0e26-44dc-b741-1cde8098e230' (osbuild) failed: error decoding osbuild output: unexpected end of JSON input\nthe raw output:\n"                                                                                                               |
| time="2022-10-25T09:33:47Z" level=error msg="osbuild job failed: osbuild build failed" jobId=77f48f72-0e26-44dc-b741-1cde8098e230                                                                                                                                                                         |
|     sys.exit(load_entry_point('osbuild==69', 'console_scripts', 'osbuild')())                                                                                                                                                                                                                             |
|   File "/usr/lib/python3.9/site-packages/osbuild/main_cli.py", line 157, in osbuild_cli                                                                                                                                                                                                                   |
|     manifest.download(object_store, monitor, args.libdir)                                                                                                                                                                                                                                                 |
|   File "/usr/lib/python3.9/site-packages/osbuild/pipeline.py", line 390, in download                                                                                                                                                                                                                      |
|     source.download(mgr, store, libdir)                                                                                                                                                                                                                                                                   |
|   File "/usr/lib/python3.9/site-packages/osbuild/sources.py", line 41, in download                                                                                                                                                                                                                        |
|     reply = client.call_with_fds("download", args, fds)                                                                                                                                                                                                                                                   |
|   File "/usr/lib/python3.9/site-packages/osbuild/host.py", line 384, in call_with_fds                                                                                                                                                                                                                     |
|     raise error                                                                                                                                                                                                                                                                                           |
| osbuild.host.RemoteError: CalledProcessError: Command '['ostree', 'pull', '61de01e4-ff03-4352-a48b-dcd91e52a1e2', '18a5cf2668ab6ab1887b9dfa82c3ae0f33e85b415cb9179bdcbdefee23baf74a', '--repo=/var/cache/osbuild-worker/osbuild-store/sources/org.osbuild.ostree/repo']' returned non-zero exit status 1. |
|    File "/usr/lib/python3.9/site-packages/osbuild/host.py", line 268, in serve                                                                                                                                                                                                                            |
|     reply, reply_fds = self._handle_message(msg, fds)                                                                                                                                                                                                                                                     |
|   File "/usr/lib/python3.9/site-packages/osbuild/host.py", line 301, in _handle_message                                                                                                                                                                                                                   |
|     ret, fds = self.dispatch(name, args, fds)                                                                                                                                                                                                                                                             |
|   File "/usr/lib/python3.9/site-packages/osbuild/sources.py", line 106, in dispatch                                                                                                                                                                                                                       |
|     self.download(SourceService.load_items(fds))                                                                                                                                                                                                                                                          |
|   File "/usr/lib/python3.9/site-packages/osbuild/sources.py", line 82, in download                                                                                                                                                                                                                        |
|     for _ in executor.map(self.fetch_one, *zip(*transformed)):                                                                                                                                                                                                                                            |
|   File "/usr/lib64/python3.9/concurrent/futures/_base.py", line 609, in result_iterator                                                                                                                                                                                                                   |
|     yield fs.pop().result()                                                                                                                                                                                                                                                                               |
|   File "/usr/lib64/python3.9/concurrent/futures/_base.py", line 446, in result                                                                                                                                                                                                                            |
|     return self.__get_result()                                                                                                                                                                                                                                                                            |
|   File "/usr/lib64/python3.9/concurrent/futures/_base.py", line 391, in __get_result                                                                                                                                                                                                                      |
|     raise self._exception                                                                                                                                                                                                                                                                                 |
|   File "/usr/lib64/python3.9/concurrent/futures/thread.py", line 58, in run                                                                                                                                                                                                                               |
|     result = self.fn(*self.args, **self.kwargs)                                                                                                                                                                                                                                                           |
|   File "/usr/lib/osbuild/sources/org.osbuild.ostree", line 107, in fetch_one                                                                                                                                                                                                                              |
|     ostree("pull", uid, commit, repo=self.repo)                                                                                                                                                                                                                                                           |
|   File "/usr/lib/osbuild/sources/org.osbuild.ostree", line 69, in ostree                                                                                                                                                                                                                                  |
|     subprocess.run(["ostree"] + args,                                                                                                                                                                                                                                                                     |
|   File "/usr/lib64/python3.9/subprocess.py", line 528, in run                                                                                                                                                                                                                                             |
|     raise CalledProcessError(retcode, process.args,                                                                                                                                                                                                                                                       |
| Traceback (most recent call last):                                                                                                                                                                                                                                                                        |
|   File "/usr/bin/osbuild", line 33, in <module>                                                                                                                                                                                                                                                           |
| libostree HTTP error from remote 61de01e4-ff03-4352-a48b-dcd91e52a1e2 for <https://microshift-kubecon22-demo.s3.us-east-2.amazonaws.com/repo/objects/0b/64c3431d9dd4a5c857b9e21edfca1d7be89b56add3aaf5bd7f72ea676401b7.filez>: Server returned HTTP 500                                                   |
| time="2022-10-25T09:32:45Z" level=info msg="Instance protected"                                                                                                                                                                                                                                           |
| time="2022-10-25T09:32:45Z" level=info msg="Running job '77f48f72-0e26-44dc-b741-1cde8098e230' (osbuild)\n"                                                                                                                                                                                               |
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by mangelajo** [comment]
> Maybe I'm missing some details, but how did you deduce this happened in reaction to an HTTP 500 response? The logs you pasted seem to point instead to `CURLE_RECV_ERROR`, which should be some kind of problem with the underlying network/transport instead, I think. At that point `libcurl` is possibly having issues with your whole networking I/O in general (i.e.possibly affecting DNS, TLS, etc.), not just with a HTTP `GET` to retry.

It's something I have seen from to sides, on @oglok 's demo with manual use of S3 buckets, and via the use Edge Management fleet manager makes of s3 buckets to distribute ostree repos to the devices for updates. 

I believe it's an issue in AWS S3, and retrying seems to fix it.

**Comment by mangelajo** [comment]
```
libostree HTTP error from remote 61de01e4-ff03-4352-a48b-dcd91e52a1e2 for <https://microshift-kubecon22-demo.s3.us-east-2.amazonaws.com/repo/objects/0b/64c3431d9dd4a5c857b9e21edfca1d7be89b56add3aaf5bd7f72ea676401b7.filez>: Server returned HTTP 500   

```

**Comment by oglok** [comment]
@cgwalters @dbnicholson should we close this issue now?

**Comment by dbnicholson** [comment]
Indeed. Fixed by #2759.
